### PR TITLE
Enable PodSecurityPolicy with privileged and unprivileged defaults

### PR DESCRIPTION
**Pull request by tvs**
**What this PR does / why we need it**:
CFCR currently does not expose a mechanism by which a cluster operator can control the security policies set on their cluster. As an example, the operator may wish to prevent all but a small subset of users from deploying privileged containers. Under the current mechanism, `allow-privileged` is an all-or-nothing flag that applies to _all_ users and service accounts.

This PR enables the [PodSecurityPolicy](https://kubernetes.io/docs/concepts/policy/pod-security-policy) access controller, which allows the operator to apply more fine-grained controls over who can and cannot access certain security features. It also disables `SecurityContextDeny` as the two aren't meant to be used at the same time and it can be modeled through `PodSecurityPolicy`.

The default pod security policies only allows members of the group `system:masters` to deploy privileged containers (through `psp/privileged`). Members of the group `system:serviceaccounts` and `system:authenticated` (that is, service accounts and authenticated users) are granted otherwise wholly-permissive privileges, but _are not_ allowed to deploy privileged containers. The constructs can
be found in the following:

* `PodSecurityPolicy`: `privileged` (`psp/privileged`) and `unprivileged` (`psp/unprivileged`)
* `ClusterRole`: `psp:privileged` (`clusterrole/psp:privileged`) and `psp:unprivileged` (`clusterrole/psp:unprivileged`)
* `ClusterRoleBinding`: `psp:privileged` (`clusterrolebinding/psp:privileged`) and `psp:unprivileged` (`clusterrolebinding/psp:unprivileged`)

Theoretically, these policies should be applied as part of a pre-upgrade hook, else service accounts and non-`cluster-admin` users will be unable to deploy pods in the period of time between the apiserver starting with the admission controller and the policies being made. Normally this period of time will be quite small and Kubernetes won't tear down any existing workloads, but if a `replicaset` is attempting to restore a pod, or something gets deployed during the upgrade, it will be unable to
start a pod until the policy is applied.

**How can this PR be verified?**
[The PodSecurityPolicy doc provides a good example](https://kubernetes.io/docs/concepts/policy/pod-security-policy#example). Running through the example to make sure that PSPs behave as-expected would require that the `ClusterRoleBinding` `psp:privileged` is deleted first. Once done, the example can be run through

To test that the `cluster-admin` can use the `PodSecurityPolicies` that we create by default, we can run these commands:

```
$ kubectl auth can-i use podsecuritypolicy/privileged
yes
$ kubectl auth can-i use podsecuritypolicy/unprivileged
yes
```

To test the privileges granted to a service account we can us these commands:

```
$ kubectl --as=system:serviceaccounts:kube-dns auth can-i use podsecuritypolicy/privileged
yes
$ kubectl --as=system:serviceaccounts:kube-dns auth can-i use podsecuritypolicy/unprivileged
no
```

**Is there any change in kubo-deployment?**
No.

**Is there any change in kubo-ci?**
No. The tests are all leveraging `kubectl` with the `cluster-admin` credentials.  It's possible that a test which performs the above `kubectl auth can-i` commands could be added.

**Does this affect upgrade, or is there any migration required?**
`SecurityContextDeny` has been removed, so anyone relying on it will need to model the same behavior in a `PodSecurityPolicy`. `psp/unprivileged` is more-permissive than `SecurityContextDeny`, so there should be no effect on running workloads.

The policy _should_ be applied before upgrade to prevent any potential downtime when it comes to deploying additional pods. Until the policy gets applied, pods will not be able to be scheduled. This period of time should be quite short as it's only between the time when the first `kube-apiserver` starts up and when `kubernetes-roles` gets run.

**Which issue(s) this PR fixes:**
[#155793102](https://www.pivotaltracker.com/story/show/155793102)

**Release note**:

```release-note
`SecurityContextDeny` is no longer enabled when privileged containers are disabled.
Instead, it has been replaced with `PodSecurityPolicy` that is always enabled and has
a default policy that is as permissive as possible. Cluster operators should build a more
strict set of policies that better secure their cluster in production.
```

### CODE REVIEW

**Comment by iainsproat** [comment]
Hi @tvs 
We're looking at this for context related to the experiment @mordebites and I are running this week.  Is the `podsecuritypolicy.yml.erb` file you've provided just used for testing only?  If so, is there a specific reason it is in the job instead of in a test directory?  If not, we don't quite see where it is applied to the cluster. Could you provide us with some explanation?

Thanks

**Comment by tvs** [comment]
@iainsproat Everything in `kubernetes-roles/config/policies` is applied by the [apply_rbac](https://github.com/cloudfoundry-incubator/kubo-release/blob/2791180507587702a46acc1bb93d990c79544752/jobs/kubernetes-roles/templates/bin/apply_rbac.erb) script.

At least one PSP must be present else Kubernetes will reject all inbound pod requests, which will prevent the `apply-specs` errand from running to completion. The PSP I've provided here is _not_ just for testing:

>The default pod security policies only allows members of the group `system:masters` to deploy privileged containers (through `psp/privileged`). Members of the group `system:serviceaccounts` and `system:authenticated` (that is, service accounts and authenticated users) are granted otherwise wholly-permissive privileges, but _are not_ allowed to deploy privileged containers. The constructs can
be found in the following:
>
>* `PodSecurityPolicy`: `privileged` (`psp/privileged`) and `unprivileged` (`psp/unprivileged`)
>* `ClusterRole`: `psp:privileged` (`clusterrole/psp:privileged`) and `psp:unprivileged` (`clusterrole/psp:unprivileged`)
>* `ClusterRoleBinding`: `psp:privileged` (`clusterrolebinding/psp:privileged`) and `psp:unprivileged` (`clusterrolebinding/psp:unprivileged`)

**Comment by obeyler** [comment]
@tvs  May we can change the  spec of job release to have a property on bosh release witch set by default the -enable-admission-plugins and another for disable-admission-plugins and let operator decide to add or remove some plugins such as PodSecurityPolicy instead of forcing user to have PodSecurityPolicy in all case
This can be helpfull to add or remove other adminssion-plugins

**Comment by tvs** [comment]
@obeyler That would be my preferred approach as well. I had a PR that was intended to allow that flexibility as well: https://github.com/cloudfoundry-incubator/kubo-release/pull/206

That said, it would need additional mechanisms by which to apply fixtures (e.g. the default PSPs) during the deployment phase, else the system containers would never be able to start. I think CFCR should follow that approach anyway - moving the various system pods, RBAC roles, etc. into the deployment manifest so that everything is configurable - but I haven't seen much to indicate we'd go that route.

**Comment by obeyler** [comment]
@tvs  I've try to discuss about that on slack channel
https://cloudfoundry.slack.com/messages/C5SF37P8X/convo/C5SF37P8X-1536048366.000100/?

File: jobs/kubernetes-roles/templates/config/policies/podsecuritypolicy.yml.erb [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: psp:unprivileged
+rules:
+- apiGroups:
+  - extensions
+  resourceNames:
+  - privileged
```
Comment (st3v) [inline jobs/kubernetes-roles/templates/config/policies/podsecuritypolicy.yml.erb]: `unprivileged` I assume

**Comment by iainsproat** [comment]
This is now under development as part of this story https://www.pivotaltracker.com/story/show/155793102
The current work builds on top of the [configurable flags feature](https://www.pivotaltracker.com/story/show/160558973).

**Comment by alex-slynko** [comment]
We made it possible to enable it without defaults in 0.24.
Yay!

**Pull request closed without merging**